calculator/calculator.py

```diff
@@ -1,15 +1,159 @@
 import re
+import math
+from statistics import mean
+from typing import Union
 
 
-class Calculator:
+__TOKEN_SPECIFICATION = {
+    'ADDITION': r'\+',
+    'SUBTRACTION': r'-',
+    'MULTIPLICATION': r'\*',
+    'DIVISION': r'/',
+    'POWER': r'\^',
+    'NEGATE': r'~',
+    'MODULE': r'%',
+    'FACTORIAL': r'!',
+    'AVERAGE': r'@',
+    'MAX': r'\$',
+    'MIN': r'&',
+    'LEFT_PARENTHESIS': r'\(',
+    'RIGHT_PARENTHESIS': r'\)',
+    'NUMBER': '\w+',
+    'NUMBER_FLOAT': '\w+\.\w+',
+}
 
-    @classmethod
-    def calculate(cls, expression=str):
-        if type(expression) is not str:
-            raise TypeError("Expression input should be of str type.")
+__NUMBER_REGEX = f'{__TOKEN_SPECIFICATION["SUBTRACTION"]}{__TOKEN_SPECIFICATION["NUMBER_FLOAT"]}|' \
+                 f'{__TOKEN_SPECIFICATION["SUBTRACTION"]}{__TOKEN_SPECIFICATION["NUMBER"]}|' \
+                 f'{__TOKEN_SPECIFICATION["NUMBER_FLOAT"]}|' \
+                 f'{__TOKEN_SPECIFICATION["NUMBER"]}'
 
-        # TODO
+__OPERATION_HANDLERS = (
+    "__handle_parenthesis",
+    "__handle_factorial",
+    "__handle_negate",
+    "__handle_max_min_average",
+    "__handle_power",
+    "__handle_multiplication_division",
+    "__handle_addition_subtraction",
+)
 
 
+def calculate(expression: str) -> Union[int, float]:
+    if type(expression) is not str:
+        raise TypeError("Expression input should be of str type.")
 
+    return __calculate(re.sub("\s", "", expression))
 
+
+def __calculate(expression: str) -> Union[int, float]:
+    if __is_end_of_calculation(expression):
+        return expression
+
+    for func in __OPERATION_HANDLERS:
+        tmp_result = eval(func)(expression)
+        if tmp_result is None:
+            continue
+        else:
+            return __calculate(tmp_result)
+
+    raise ValueError(f'Expression: {expression} is not legal!')
+
+
+def __is_end_of_calculation(expression: str) -> bool:
+    if re.search(f'^({__NUMBER_REGEX})$', expression):
+        return True
+
+    return False
```

Comment:
This function is doing one line, just inline the function in the condition above (in the _calculate function)